### PR TITLE
[SPARK-13386][GraphX] ConnectedComponents should support maxIteration option

### PR DESCRIPTION
**Pull request by zhengruifeng**
JIRA: https://issues.apache.org/jira/browse/SPARK-13386
## What changes were proposed in this pull request?

add maxIteration option for ConnectedComponents algorithm
## How was the this patch tested?

unit tests passed

### CODE REVIEW

**Comment by srowen** [comment]
Although that's probably fine, it does change a public API. I'd prefer a maintainer look at it, but I'm not sure who is still available.

File: graphx/src/main/scala/org/apache/spark/graphx/GraphOps.scala [hunk 1]
```diff
@@ -411,8 +411,8 @@ class GraphOps[VD: ClassTag, ED: ClassTag](graph: Graph[VD, ED]) extends Seriali
    *
    * @see [[org.apache.spark.graphx.lib.ConnectedComponents$#run]]
    */
-  def connectedComponents(): Graph[VertexId, ED] = {
-    ConnectedComponents.run(graph)
+  def connectedComponents(numIter: Int = Int.MaxValue): Graph[VertexId, ED] = {
```
Comment (rxin) [inline graphx/src/main/scala/org/apache/spark/graphx/GraphOps.scala]: we should just add a new method to this.

Comment (zhengruifeng) [inline graphx/src/main/scala/org/apache/spark/graphx/GraphOps.scala]: you are right, I have fixed it. Now there is another ConnectedComponents API with numIters

File: graphx/src/main/scala/org/apache/spark/graphx/GraphOps.scala [hunk 2]
```diff
@@ -406,13 +406,23 @@ class GraphOps[VD: ClassTag, ED: ClassTag](graph: Graph[VD, ED]) extends Seriali
...
+
+  /**
    * Compute the connected component membership of each vertex and return a graph with the vertex
    * value containing the lowest vertex id in the connected component containing that vertex.
    *
    * @see [[org.apache.spark.graphx.lib.ConnectedComponents$#run]]
    */
-  def connectedComponents(): Graph[VertexId, ED] = {
-    ConnectedComponents.run(graph)
+  def connectedComponents(numIter: Int): Graph[VertexId, ED] = {
```
Comment (rxin) [inline graphx/src/main/scala/org/apache/spark/graphx/GraphOps.scala]: maybe name this maxIterations to be consistent with Pregel's parameter.

File: graphx/src/main/scala/org/apache/spark/graphx/lib/ConnectedComponents.scala [hunk 3]
```diff
@@ -29,13 +29,14 @@ object ConnectedComponents {
...
-   *
    * @param graph the graph for which to compute the connected components
-   *
+   * @param numIter the maximum number of iterations to run for
    * @return a graph with vertex attributes containing the smallest vertex in each
    *         connected component
    */
-  def run[VD: ClassTag, ED: ClassTag](graph: Graph[VD, ED]): Graph[VertexId, ED] = {
+  def run[VD: ClassTag, ED: ClassTag](graph: Graph[VD, ED],
+                                      numIter: Int = Int.MaxValue): Graph[VertexId, ED] = {
```
Comment (rxin) [inline graphx/src/main/scala/org/apache/spark/graphx/lib/ConnectedComponents.scala]: we should do the same thing here too to avoid breaking apis.

Comment (rxin) [inline graphx/src/main/scala/org/apache/spark/graphx/lib/ConnectedComponents.scala]: and call it maxIterations

Comment (zhengruifeng) [inline graphx/src/main/scala/org/apache/spark/graphx/lib/ConnectedComponents.scala]: ok, all those options are renamed

File: graphx/src/main/scala/org/apache/spark/graphx/lib/ConnectedComponents.scala [hunk 4]
```diff
@@ -29,13 +29,14 @@ object ConnectedComponents {
...
-   *
    * @param graph the graph for which to compute the connected components
-   *
+   * @param maxIterations the maximum number of iterations to run for
    * @return a graph with vertex attributes containing the smallest vertex in each
    *         connected component
    */
-  def run[VD: ClassTag, ED: ClassTag](graph: Graph[VD, ED]): Graph[VertexId, ED] = {
+  def run[VD: ClassTag, ED: ClassTag](graph: Graph[VD, ED],
+                                      maxIterations: Int = Int.MaxValue): Graph[VertexId, ED] = {
```
Comment (rxin) [inline graphx/src/main/scala/org/apache/spark/graphx/lib/ConnectedComponents.scala]: for this one, we should also just overload it so we don't break binary compatibility.

Comment (zhengruifeng) [inline graphx/src/main/scala/org/apache/spark/graphx/lib/ConnectedComponents.scala]: ok, another run api without maxIterations options is added

**Pull request closed without merging**